otx/algorithms/segmentation/adapters/mmseg/models/segmentors/mean_teacher_segmentor.py

```diff
@@ -61,37 +129,61 @@ def forward_dummy(self, img, **kwargs):
     def forward_train(self, img, img_metas, gt_semantic_seg, **kwargs):
         """Forward train."""
         self.count_iter += 1
-        if self.semisl_start_iter > self.count_iter or "extra_0" not in kwargs:
+        self.losses["sum_loss"] = 0.0
+        if self.semisl_start_iter >= self.count_iter or "extra_0" not in kwargs:
             x = self.model_s.extract_feat(img)
             loss_decode = self.model_s._decode_head_forward_train(x, img_metas, gt_semantic_seg=gt_semantic_seg)
-            return loss_decode
+            self._update_summary_loss(loss_decode)
+            if self.use_prototype_head:
+                self.decode_proto_network(x, gt_semantic_seg)
+
+            # add information about accuracy
+            for key in loss_decode:
+                if "acc" in key and loss_decode[key] is not None:
+                    self.losses["decode_acc"] = loss_decode[key]
 
+            return self.losses
+
+        # + unsupervised part
         ul_data = kwargs["extra_0"]
-        ul_s_img = ul_data["img"]
-        ul_w_img = ul_data["ul_w_img"]
+        ul_s_img = ul_data["img"]  # strongly augmented
+        ul_w_img = ul_data["ul_w_img"]  # weakly augmented
         ul_img_metas = ul_data["img_metas"]
 
-        with torch.no_grad():
-            teacher_feat = self.model_t.extract_feat(ul_w_img)
-            teacher_logit = self.model_t._decode_head_forward_test(teacher_feat, ul_img_metas)
-            teacher_logit = resize(
-                input=teacher_logit, size=ul_w_img.shape[2:], mode="bilinear", align_corners=self.align_corners
-            )
-            _, pl_from_teacher = torch.max(torch.softmax(teacher_logit, axis=1), axis=1, keepdim=True)
-
-        losses = dict()
+        # generate pseudo labels, filter high entropy pixels, compute loss reweight
+        pl_from_teacher, reweight_unsup = self.generate_pseudo_labels(ul_w_img, ul_img_metas)
 
+        # extract features from labeled and unlabeled augmented images
         x = self.model_s.extract_feat(img)
         x_u = self.model_s.extract_feat(ul_s_img)
         loss_decode = self.model_s._decode_head_forward_train(x, img_metas, gt_semantic_seg=gt_semantic_seg)
         loss_decode_u = self.model_s._decode_head_forward_train(x_u, ul_img_metas, gt_semantic_seg=pl_from_teacher)
-
-        for (key, value) in loss_decode_u.items():
-            if value is None:
+        self._update_summary_loss(loss_decode)
+        self._update_summary_loss(loss_decode_u, loss_weight=self.unsup_weight * reweight_unsup)
+
+        if self.use_prototype_head:
+            # for proto head we need to derive head features
+            self.decode_proto_network(x, gt_semantic_seg, x_u, pl_from_teacher, reweight_unsup)
+
+        if hasattr(self.model_s, "auxiliary_head"):
```

Comment:
Do we have a case to keep auxiliary head in OTX?

Reply:
Some methods, for example, deeplabv3 ResNet family supports Aux head. 
Actually, I conducted experiments with that for HRNets and SegNext and saw the perfomance drop. So, I think we can avoid keeping AUX head.